### ISSUE TITLE
ARM builds?

### ISSUE THREAD

**Issue by takluyver** [pypa/manylinux#84]
This is a very speculative issue. I'm aware that PEP 513 says manylinux1 is only defined for i686 and x86_64. But it would be nice if instructions telling people to `pip install` compiled modules worked easily and quickly on things like the Raspberry Pi too. It might also be valuable to the few brave souls trying to make Python work on mobile platforms.

So this issue is a place to work out what would be needed.

1. **Architecture requirements**: there are a lot of different versions of ARM, and there are confusingly similar looking version schemes for the architectures (e.g. `ARMv7`) and cores (e.g. `ARM11`). There may also be optional modules. Debian supports three buckets of ARM architectures: `armhf` requires ARMv7 and a floating point unit (hf = hardware float). `armel` supports older architectures, and cores without an FPU. `arm64` is for the newer 64-bit processors. In terms of raspis:

- Raspi 0 and 1 have ARMv6 architecture, so Debian armhf does not support them. They do have FPUs, though, so Raspian is compiled to make use of this.
- Raspi 2 has ARMv7
- Raspi 3 has a 64-bit ARMv8 architecture.

2. **Minimum library versions**: These probably don't need to be all that old, because the systems I think this would target are likely to be running a relatively recent distro. But we'd need to work out what they are.

3. **Distro for building**: What distro would be the equivalent of the Centos 5.11 we use for building x86 and x64 manylinux wheels?

4. **Build platform**: Are there any CI services which offer ARM builds? Is virtualisation fast enough? Can we make it easy for people to build wheels on their own Raspberry Pi or similar device.

**Comment by rmcgibbo** [comment]
I am not personally very knowledgeable about ARM, so I don't know the direct answers to any of those questions. It would be really excellent if manylinux could support ARM at some point in the future though -- I have the impression that it's a growing platform.

For the manylinux1 spec, @njsmith and I leveraged a lot of Continuum's work to define the base distro and minimum library versions. Since the manylinux1 release, Continuum has added [some support](https://www.continuum.io/content/conda-support-raspberry-pi-2-and-power8-le) for ARM to conda and their packages, so perhaps this would be a good place to begin to identify best practices.

**Comment by takluyver** [comment]
Thanks, that's good to know. It looks like they're targeting ARMv7 and using Raspbian to run the builds.

**Comment by njsmith** [comment]
Unless someone wants to do lots of abi compatibility testing across
different arm distributions, I think the most viable way forward here would
be to push forward with the idea of distro-specific wheel tags, so that
people could upload e.g "raspbian wheels"?

**Comment by KellenSunderland** [comment]
Using dockcross works quite well for ARM builds.  I've created a few packages that use cross-compilation rather than virtualization.  Builds that take hours on a Pi typically take a minute or two.  This Dockerfile for example builds packages that run well for me with a Pi3 + Raspbian jessi.  https://github.com/dockcross/dockcross/blob/master/linux-armv7/Dockerfile

**Comment by takluyver** [comment]
Since my motivation for this was getting Python packages on a Raspberry Pi, I'll drop a link to [piwheels](https://www.piwheels.hostedpi.com/), @bennuttall's project to compile wheels for that platform.

**Comment by bennuttall** [comment]
An overview:

Although the hardware spec of Pi 1/Zero is ARMv6, Pi 2 is ARMv7 and Pi 3 is ARMv8, the reality is that Raspbian runs in ARMv6 user space regardless on your hardware. When you build a wheel on a Pi it's tagged armv6l on Pi 1/Zero and armv7l on Pi 2/3, but really it's always ARMv6.

So my project [piwheels](https://www.piwheels.hostedpi.com/) provides wheels for Raspberry Pi, but I wouldn't bank on the platform wheels working on other ARM boards. There's no particular need for maintainers to upload their wheels to PyPI as Pi users can use piwheels as their repo instead. Soon, in Raspbian, pip will be configured to use piwheels.

**Comment by QJKX** [comment]
[Python on Termux](https://wiki.termux.com/wiki/Python) (an android Linux environment) would also benefit from arm wheels.

**Comment by NotAFile** [comment]
We're having this problem with testing ARM in CI. The lack of wheels forces us to compile all dependencies for a fresh build, which takes around 20 minutes.

**Comment by vielmetti** [comment]
There is a CI system available, Drone Cloud, that can build natively for 32-bit and 64-bit Arm systems. There's also bare metal resources in the @worksonarm project to tackle this. Happy to help tackle this, especially for arm64 (aarch64).

**Comment by vielmetti** [comment]
@bennuttall - does piwheels support 32-bit Arm only, or do you also have 64-bit arm64 support?

**Comment by bennuttall** [comment]
We only provide wheels for 32-bit systems.

**Comment by vielmetti** [comment]
Referencing https://marcin.juszkiewicz.com.pl/2019/02/23/we-need-arm64-systems-for-developers-again/ which describes the resulting pain that `pip install scipy`, which takes a few seconds on x86, takes hours on arm64 because of lack of wheels.

**Comment by hrw** [comment]
I discussed that with Pypi devs some time ago on irc.

Whole service depends on whatever authors upload. So if we want non-x86 wheels then we need to convince authors to build and upload arm wheels.

**Comment by hrw** [comment]
And when it comes to what to use as a minimal...

I think that Debian jessie is sensible minimum. Eventually CentOS 7.

Older releases did not supported arm64.

**Comment by ralphie-boy** [comment]
I need to do python development that can include the python SpeechRecognition package. Samsung recently released the Galaxy Book 2 with the latest 64 bit Qualcomm Snapdragon processor (850). It comes with Windows 10 S which can be upgraded to Windows Pro. Windows Pro will support WSL (Windows Subsystem for Linux). WSL will allow me seamlessly run Ubuntu. Ubuntu will have built-in Python. The question is what happens when I try to install the SpeechRecognition package? I know for example this will not work on iPad. Will the Arm 64 architecture support this package?

**Comment by takluyver** [comment]
SpeechRecognition itself appears to be a pure Python package, so there should be no problem installing it anywhere Python works. But it looks like it's a wrapper around various local and remote speech recognition engines, so the actual difficulty will probably depend which backend you need. But it's almost certainly possible - you just might need to install some things through apt or compile them from source.

**Comment by ralphie-boy** [comment]
Thanks for the reply, takluyver. I think you are correct. But I failed to procure the Galaxy Book 2 in time for my trip. Instead I'm taking my iPad, on which, Pythonista is the only supported Python. So SpeechRecognition won't work, because Pythonista is not a true complete working Python. It's pretty good, but it is limited to a subset of packages for which dependencies can be supported by whatever wrappers.

**Comment by ulope** [comment]
On a slightly related note docker [has supported multi-platform images for a while now through manifest lists](https://blog.docker.com/2017/11/multi-arch-all-the-things/) and all official (i.e. made by docker) images on docker hub are multi platform.

Together with the new [buildx multi platform support](https://engineering.docker.com/2019/04/multi-arch-images/) this could make providing ARM images much easer.

**Comment by takluyver** [comment]
Just to remind everyone who might contribute here: the essential first step would be to define appropriate ARM architecture tags. What instruction sets & extensions are relevant, what are the rules for compatibility (e.g. presumably a binary built with NEON extensions won't work on a CPU that doesn't have them), and how can we detect what the hardware supports?

If we can work that out, I'd be optimistic that we could do something for glibc-based distros on ARM relatively easily, based on what's already been done for x86 & x86_64.

An alternative approach would be to focus on defining more specific tags, e.g. for raspbian. But that's obviously not as broadly useful.

**Comment by hrw** [comment]
@takluyver for ARMv7 it should follow 'armhf' defined by Debian/Fedora/etc distros. No NEON, -vfp3d16.

AArch64 (aka arm64) is simpler - just gcc defaults.

And target Debian/jessie or Debian/stretch as minimal OS version and binaries should (imho) work under CentOS 7/8, Debian 8+, Ubuntu 16.04+ etc.

**Comment by lpsinger** [comment]
It would be more convenient if the arm manylinux image was based on the same OS as the x86_64 image. Some times I need to install additional build tools and it would create some friction if I had to use yum for x86_64 but apt for arm. Is there a CentOS port for ARM?

**Comment by lpsinger** [comment]
Also I take it that an armhf does not include NEON but aarch64 does?

**Comment by mattip** [comment]
Correct me if I'm wrong. NEON and friends are not part of the spec, just like AVX is not part of the x64 spec. It is up to c-extension library writers to deal with detecting CPU capabilities and dealing with their presence or absence at run-time. NumPy does this by having separate loops for different architecture variants, as does Intel in their various packages.

**Comment by lpsinger** [comment]
> Correct me if I'm wrong. NEON and friends are not part of the spec, just like AVX is not part of the x64 spec.

This is unclear to me. According to https://en.wikipedia.org/wiki/ARM_architecture#ARMv8-A:

> To both AArch32 and AArch64, ARMv8-A makes VFPv3/v4 and advanced SIMD (NEON) standard.

**Comment by hrw** [comment]
> It would be more convenient if the arm manylinux image was based on the same OS as the x86_64 image. Some times I need to install additional build tools and it would create some friction if I had to use yum for x86_64 but apt for arm. Is there a CentOS port for ARM?

CentOS 7 supports aarch64 as one of altarch architetures. 32bit arm port is available but with other level of support iirc.

**Comment by mayeut** [comment]
To sum-up a few comments:
> Also I take it that an armhf does not include NEON but aarch64 does?

Yes

> Correct me if I'm wrong. NEON and friends are not part of the spec, just like AVX is not part of the x64 spec.

> This is unclear to me. According to https://en.wikipedia.org/wiki/ARM_architecture#ARMv8-A:
> To both AArch32 and AArch64, ARMv8-A makes VFPv3/v4 and advanced SIMD (NEON) standard.

To extend @hrw  https://github.com/pypa/manylinux/issues/84#issuecomment-531679045:
> ARMv7 it should follow 'armhf' defined by Debian/Fedora/etc distros. No NEON, -vfp3d16.

armhf: 32bits, minimum target armv7, No NEON, -vfp3d16
> AArch64 (aka arm64) is simpler - just gcc defaults.

aarch64: 64bits, minimum target armv8, NEON by default

**Comment by lpsinger** [comment]
> > It would be more convenient if the arm manylinux image was based on the same OS as the x86_64 image. Some times I need to install additional build tools and it would create some friction if I had to use yum for x86_64 but apt for arm. Is there a CentOS port for ARM?
> 
> CentOS 7 supports aarch64 as one of altarch architetures. 32bit arm port is available but with other level of support iirc.

How relevant is 32-bit armhf support these days? If you only supported aarch64, would that be a problem?

If armhf is important, is the CentOS 7 support for it good enough? If not, is the Fedora picture any better?

**Comment by mattip** [comment]
I would vote for starting with 64 bit, and leaving armhf out for now.

**Comment by bennuttall** [comment]
> How relevant is 32-bit armhf support these days? If you only supported aarch64, would that be a problem?

The Raspberry Pi generally runs a 32-bit operating system. Pi 3 and 4 are 64-bit CPUs but the Raspbian image runs in 32-bit mode to support all boards. Plus Pi 1 and Pi Zero are still sold, and they are armv6.

**Comment by lpsinger** [comment]
Looks like the CentOS docker images support armhf and aarch64 as of CentOS 7 and later (https://hub.docker.com/r/arm32v7/centos/, https://hub.docker.com/r/arm64v8/centos/), so ARM support based upon them would be possible as of manylinux2014 (but not manylinux2010).

**Comment by lpsinger** [comment]
So this would depend on #348.

**Comment by mayeut** [comment]
If someone has access to an aarch64 linux platform with docker, it might be worth trying to build https://github.com/mayeut/manylinux/tree/manylinux2014-aarch64 , it's taking ages with qemu so I don't know if I'm going to wait that long. (and debugging in case it fails is just a no go with qemu for me). Good news for aarch64 is that devtoolset-8 already exists. Not the case for armhf (where, anyway, yum segfaults qemu).
Are there any free (for OpenSource) services like travis-ci using arm based devices ? (seems like a good thread to ask that)

**Comment by hrw** [comment]
@mayeut started build. Will share on hub.docker.com once built.

When it comes to aarch64 boxes for CI then it depends. At Linaro we provided some boxes for external projects. https://www.packet.com/ and AWS are other options (both paid rather).

**Comment by hrw** [comment]
@mayeut  "haerwu/manylinux2014_aarch64:latest" pushed to Docker hub.

**Comment by mattip** [comment]
NumPy is using [shippable](https://app.shippable.com/) for aarch64 instances which has proved quite reliable.

**Comment by JonasVautherin** [comment]
Out of curiosity, is there any plan in supporting armv7 in manylinux2014? I'm not sure what it implies, I just observe that armv7l and ppc64 are listed in PEP599 but there are no manylinux2014 images for them.

**Comment by mayeut** [comment]
There's no plan to support `armv7` in `manylinux2014`. CentOS does not support this architecture.
There's a chance to get `armv7` with `PEP600` if we get a `debian`/`ubuntu` as base image.
This will still require CI service support though.
There are no official images for `ppc64` in docker hub so I guess there will be no `manylinux` image for this architecture (and it would require support from a CI service as well)

**Comment by ax3l** [comment]
Although not from the `CentOS` channel, I just found official centos images for `ppc64le` and `armv7`:
- https://github.com/docker-library/official-images#architectures-other-than-amd64
- https://hub.docker.com/r/ppc64le/centos
- https://hub.docker.com/r/arm32v7/centos

that one could try to use as a base.

**Comment by mayeut** [comment]
Thanks @ax3l,
I did not remember that centos 7 provided an armhf build.
yum is segfaulting in qemu so I can't test this.
There will also be an issue with the toolchain because devtoolset-9 is not available.
That's also the case for i686 but I'm almost there for this one.

**Comment by mattip** [comment]
TL;DR: please think twice before releasing an image for a platform that has no known open-source CI support. Official project images on PyPI are held to a higher standard than those on `piwheels.org` or even on Anaconda, and without an efficient CI solution many projects will be pressured into ad-hoc solutions that do not scale.

If the manylinux project releases official images for a platform, there is a hidden assumption that is because the community expects library authors to support it and use it to upload images to PyPI. Releasing an arm32 image without a suitable CI solution that is readily available for open-source projects, and without a plan for such a solution, will create expectations that library providers cannot be expected to fulfil. I think the PyPA organization, together with the PSF, should first ensure there is a CI solution before releasing images for new platforms. Building and testing a project such as SciPy or Astropy on a qemu-emulated machine will be very painful, and increases the complexity of debugging CI failures.

This pressure is already forcing the NumPy project to spend time and resources massaging its CI and code to adapt to the buggy aarch64 and ppc64le manylinux2014 platforms. CI sometimes fails on the resource-strained s390x offering, and NumPy has no recourse but to ignore such failures without knowing if it is due to a new feature or due to a flaky machine. As a single project we have little leverage with IBM or travis (or other CI providers) to improve the CI situation: perhaps there should be some guidance from PyPA or the PSF how to improve the situation. Now multiply NumPy's problems over the entire scientific stack which is largly based on c-extension modules, and you can see this small change in manylinux's repo could have far-reaching effects.

manylinux2014 is based on glibc 2.17 which was released [in Dec 2012](https://sourceware.org/glibc/wiki/Glibc%20Timeline). The expectation that this is a viable platform for x86 is stretching the truth, and in my opinion is unreasonable for arm64 and arm32. At the very least, any arm images should be based on a glibc that was released after that hardware spec was finalized, not before.

**Comment by ax3l** [comment]
> If the manylinux project releases official images for a platform, there is a hidden assumption that is because the community expects library authors to support it and use it to upload images to PyPI.

Hm, I understand the pain but I am not sure I can follow the "signaling/assumption" reasoning. Wouldn't manylinux-related PEPs, such as [defining platform tags for big endian `ppc64`](https://www.python.org/dev/peps/pep-0599/) or even allowed uploads to `PyPI` for exotic architectures, be signaling what you mention to developers?

I think we should define tags for [even more architectures and micro-architectures](https://archspec.readthedocs.io) and signal to the community via PyPI support/rejects in uploads. It's not bad to have more compatible manylinux base images, e.g. to drive a sub-community specific index or local builds.

**Comment by mattip** [comment]
>  It's not bad to have more compatible manylinux base images

Your position makes sense for pypa/manylinux as long as the people involved here have the bandwidth to continue to maintain the images. Help with gh-501, gh-542 and pypa/packaging#293 would be welcome.

I guess we (NumPy) will have to discuss internally how we respond to requests for additional wheels.

**Comment by ShadowJonathan** [comment]
This may be a digression, but are the changes that're suggested in https://github.com/pypa/warehouse/pull/7853 accurate to pep 600 and the projected `arm` arches to be supported?

**Comment by mayeut** [comment]
@ShadowJonathan,

warehouse maintainers requested that all architectures supported by [PEP599](https://www.python.org/dev/peps/pep-0599/) could be uploaded in pypa/warehouse#6684

pypa/warehouse#7853 only provides no-regression for those architectures and that doesn't mean there will ever be an `armv7l` image.

As @mattip stated, the lack of known open-source CI support is a major show stopper. Even with travis-ci supporting `aarch64`, `ppc64le` and `s390x`, we're seeing various issues on those platforms and, for the time being, I don't foresee expanding the supported architectures and, not before we have something acceptable working for [PEP600](https://www.python.org/dev/peps/pep-0600/).

@ax3l,

Support for even more architectures is allowed by [PEP600](https://www.python.org/dev/peps/pep-0600/). It's up to every project to allow anything.
As for manylinux base images, the previous comment on open-source CI support applies. There's also the the issue of toolchain support across architectures to be able to provide a single, consistent software environment for all architectures. This already became a hassle with `i686` on `CentOS` with `devtoolset` dropping support of `i686` and not being able to cross-compil for `i686`from `x86_64` (that would also have been a hassle but perhaps less).

As for micro-architectures, this is really outside of the scope of manylinux and is a much more generic tagging issue applying to windows, linux, macOS, ...
My 2 cents anyway (not as a maintainer of manylinux) is that I'd rather see libraries relying on micro-architecture optimizations to provide proper dynamic dispatch mechanisms with fallbacks as is the case for example with OpenBLAS or libjpeg-turbo.

**Comment by sandys** [comment]
hey guys, 
we would very much like to see ARM builds happen because of the general availability of AWS Graviton 2 instances on Amazon Cloud. 
At this point, they are best ROI for a cloud dollar. https://aws.amazon.com/ec2/graviton/  and https://www.anandtech.com/show/15578/cloud-clash-amazon-graviton2-arm-against-intel-and-amd/10

Being able to have our stack natively on ARM would cut down costs immensely. And I think, this would apply to the majority of people out there who use AWS. 
Also, given Apple is going fully ARM, we see this trend accelerating and not slowing.

**Comment by lpsinger** [comment]
I very much agree. However, note that manylinux would serve the AWS use case but not the Apple one (same CPU architecture, different OS and executable format).

**Comment by takluyver** [comment]
To be clear, you don't need *manylinux* support to build and use Python extensions for a particular ARM platform. https://www.piwheels.org/ already does this for Raspberry Pi, for example, and you could do something similar for ARM on AWS. So long as you know a particular target platform, you can compile on the same platform.

Manylinux is meant to make it possible to build a wheel on Fedora with one ARM chip and then install it on Ubuntu with a different one - which is obviously useful, but also much harder.

**Comment by geoffreyblake** [comment]
This issue is pretty old and hasn't seen updates for ~6 months, but is there still debate on whether manylinux2014_aarch64 is stable and abiding by PEP 599?  

I've set up a tester to automatically try some of the top wheels to see how Arm64 support is coming along using Graviton systems and the wheels built with manylinux2014_aarch64 appear to work well on both armv8.2 and armv8.0 systems running a cross-section of distros (Ubuntu 20.04, AL2, Centos8): https://github.com/geoffreyblake/arm64-python-wheel-tester

The failures seen by the tester for the packages tested are due to lack of wheels.

**Comment by mayeut** [comment]
Original post from @webknjaz in #1029 
> armv7l (that a lot of raspberries use because the official 64-bit kernel is still in beta) and ppc64 are present in the PEP but the images for these aren't built.
> 
> I haven't found any discussion mentioning them so I just wanted to have someplace documenting why they are missing.
> 
> My wild guess is that it's because Travis CI doesn't provide them. Any other reasons? It looks like it should be possible to build them with QEMU under GHA, no?

**Comment by mayeut** [comment]
@webknjaz,

Just to go a bit further, PEP600 which supersedes PEP599 doesn't restrict architectures at all.

The main reasons not to add those are:
- the lack of a base image (for `ppc64`)
- the lack of native CI
- debugging issues locally

QEMU does not solve everything. It's painfully slow and GHA builds are timing out now and then on `ppc64le` since I could not get them running back on travis-ci yet.
I've seen in the past and I'm still seeing crashes, only when running QEMU which prohibits me from debugging at all locally. I unfortunately do not have time to go through debugging/reporting/contributing/waiting for fixes in QEMU.

**Comment by webknjaz** [comment]
> GHA builds are timing out now and then on `ppc64le`

Do you know you could increase the timeout per-job and per-step? https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstepstimeout-minutes

**Comment by MrNaif2018** [comment]
@mayeut What about using qemu on aarch64 machines? CircleCI has added arm64 machines support and it's pretty stable, and qemu emulation runs almost at the same speed as native builds (because I guess arm32 is supported by kernel or the spec), and they might add arm32 machines soon.
As for testing locally: I agree, testing locally on amd64 with qemu is painful, but what about testing on aarch64 machines?
Oracle cloud (as part with partnership with arm I think) provides always free 24 gb ram and 4 cpu arm machine: 
https://docs.oracle.com/en-us/iaas/Content/FreeTier/freetier_topic-Always_Free_Resources.htm
So it's not fully local debugging, but pretty convenient (vscode can run via ssh too)
Would that help moving forward the armv7 wheels creation?

**Comment by Red-M** [comment]
I've just added such support for my own packages on arm64/aarch64 via using a full KVM to run the aarch64 docker containers for `manylinux` and its super slow. If you really want these, buy a bunch of RPi4s and run CI VMs on those via libvirt or KVM (hint, pimox). Otherwise piwheels exists, so make sure you provide an `sdist`.

**Comment by mayeut** [comment]
> What would be nice is providing buildchains in the manylinux containers for other cpu architectures to allow for easy building without the need for actual hardware and to provide already cross-compiled python versions that already exist for the x86 binaries (eg, drop/supporting a version of python brings it to all supported architectures). This will make the containers larger but at least building wheels (eg, for arm) won't suck as much to do so.

These do exists in some form outside of manylinux https://github.com/pypa/manylinux/issues/1125#issuecomment-871189428 however, while it might be suitable for maturin / pyO3 where there is probably a decent/recent rust cross-compiler, it's not the case for the gcc toolchain as noted in that comment.

Moving this to discussion.
